p2p/dial_test.go

```diff
@@ -0,0 +1,325 @@
+package p2p
+
+import (
+	"encoding/binary"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+)
+
+func init() {
+	spew.Config.Indent = "\t"
+}
+
+type dialtest struct {
+	init   *dialstate // state before and after the test.
+	rounds []round
+}
+
+type round struct {
+	peers []*Peer // current peer set
+	done  []task  // tasks that got done this round
+
+	new []task // the result must match this one
+}
+
+func runDialTest(t *testing.T, test dialtest) {
+	var (
+		vtime   time.Time
+		running []task
+	)
+	pm := func(ps []*Peer) map[discover.NodeID]*Peer {
+		m := make(map[discover.NodeID]*Peer)
+		for _, p := range ps {
+			m[p.rw.id] = p
+		}
+		return m
+	}
+	deltask := func(task task) {
+		for i := range running {
+			// The tests use DeepEqual to match tasks while Server
+			// uses pointer equality in order to make the tests more
+			// readable (otherwise we'd have to track them in variables).
+			_, islookup1 := running[i].(*discoverTask)
+			_, islookup2 := task.(*discoverTask)
+			if reflect.DeepEqual(running[i], task) || islookup1 && islookup2 {
+				running = append(running[:i], running[i+1:]...)
+				return
+			}
+		}
+	}
+	for i, round := range test.rounds {
+		for _, task := range round.done {
+			deltask(task)
+			test.init.taskDone(task, vtime)
+		}
+		new := test.init.newTasks(running, pm(round.peers), vtime)
+		if !sametasks(new, round.new) {
+			t.Errorf("round %d: new tasks mismatch:\ngot %v\nwant %v\nstate: %v\nrunning: %v\n",
+				i, spew.Sdump(new), spew.Sdump(round.new), spew.Sdump(test.init), spew.Sdump(running))
+		}
+		running = append(running, new...)
+		// Time advances by 16 seconds on every round.
+		vtime = vtime.Add(16 * time.Second)
+	}
+}
+
+// This test checks that dynamic dials are launched from discovery results.
+func TestDialStateDynDial(t *testing.T) {
+	runDialTest(t, dialtest{
+		init: newDialState(nil, 5),
+		rounds: []round{
+			// A discovery query is launched.
+			{
+				peers: []*Peer{
+					{rw: &conn{flags: staticDialedConn, id: uintID(0)}},
+					{rw: &conn{flags: dynDialedConn, id: uintID(1)}},
+					{rw: &conn{flags: dynDialedConn, id: uintID(2)}},
+				},
+				new: []task{&discoverTask{}},
+			},
+			// Dynamic dials are launched when it completes.
+			{
+				peers: []*Peer{
```

Comment:
I added these rather quickly to verify that static dials don't mess with `needDynDials`.
Seems I forgot this case.